New module postgresql_facts: Gathers facts about PostgreSQL servers.

Andersson007 opened the pull request:
##### SUMMARY
New module postgresql_facts: Gathers facts about PostgreSQL servers.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
postgresql_facts

##### MAIN PARAMETERS
```yaml
  filter:
    description:
      - Limit collected facts by comma separated string or YAML list.
      - Allowable values are C(version), C(databases), C(settings), C(tablespaces), C(roles),
        C(replications), C(repl_slots).
      - By default, collects all subsets.
      - You can use shell-style (fnmatch) wildcard to pass groups of values (see Examples).
      - You can use '!' before value (for example, C(!settings)) to exclude it from facts.
      - If you pass including and excluding values to the filter, for example, I(filter=!settings,ver),
        the excluding values will be ignored.
```
##### EXAMPLES
```yaml
# Display facts from postgres hosts.
# ansible postgres -m postgresql_facts

# Display only databases and roles facts from all hosts using shell-style wildcards:
# ansible all -m postgresql_facts -a 'filter=dat*,rol*'

# Display only replications and repl_slots facts from standby hosts using shell-style wildcards:
# ansible standby -m postgresql_facts -a 'filter=repl*'

# Display all facts from databases hosts except settings:
# ansible databases -m postgresql_facts -a 'filter=!settings'

- name: Collect PostgreSQL version and extensions
  postgresql_facts:
    filter: ver*,ext*

- name: Collect all subsets, excluding settings and roles
  postgresql_facts:
    filter: "!settings,!roles"

- name: Collect tablespaces and repl_slots facts
  postgresql_facts:
    filter:
      - tablesp*
      - repl_sl*

- name: Collect all facts except settings
  postgresql_facts:
    filter:
      - "!settings"
```
Part of possible returned facts (specially limited):
```json
postgresql_facts": {
        "version": {
            "major": 10, 
            "minor": 6
        },
        "tablespaces": {}, 
        "databases": {
            "mydb": {
                "access_priv": "", 
                "collate": "en_US.UTF-8", 
                "ctype": "en_US.UTF-8", 
                "encoding": "UTF8", 
                "owner": "mydbuser", 
                "size": "45 MB"
                "extensions": {
                    "hstore": {
                        "description": "data type for storing sets of (key, value) pairs", 
                        "extversion": {
                            "major": 1, 
                            "minor": 5
                        }, 
                        "nspname": "public"
                    }, 
                    "plpgsql": {
                        "description": "PL/pgSQL procedural language", 
                        "extversion": {
                            "major": 1, 
                            "minor": 0
                        }, 
                        "nspname": "pg_catalog"
                    }
                }, 
                "languages": {
                    "c": {
                        "lanacl": "", 
                        "lanowner": "postgres"
                    }, 
                    "internal": {
                        "lanacl": "", 
                        "lanowner": "postgres"
                    }, 
                    "plpgsql": {
                        "lanacl": "", 
                        "lanowner": "postgres"
                    }, 
                    "sql": {
                        "lanacl": "", 
                        "lanowner": "postgres"
                    }
                }, 
                "namespaces": {
                    "information_schema": {
                        "nspacl": "{postgres=UC/postgres,=U/postgres}", 
                        "nspowner": "postgres"
                    }, 
                    "pg_catalog": {
                        "nspacl": "{postgres=UC/postgres,=U/postgres}", 
                        "nspowner": "postgres"
                    }, 
                    "pg_temp_1": {
                        "nspacl": "", 
                        "nspowner": "postgres"
                    }, 
                    "pg_toast": {
                        "nspacl": "", 
                        "nspowner": "postgres"
                    }, 
                    "pg_toast_temp_1": {
                        "nspacl": "", 
                        "nspowner": "postgres"
                    }, 
                    "public": {
                        "nspacl": "{postgres=UC/postgres,=UC/postgres}", 
                        "nspowner": "postgres"
                    }
                }, 
            }, 
            "template0": {
                "access_priv": "=c/postgres\npostgres=CTc/postgres", 
                "collate": "en_US.UTF-8", 
                "ctype": "en_US.UTF-8", 
                "encoding": "UTF8", 
                "owner": "postgres", 
                "size": "7729 kB"
            }
        }, 
        "repl_slots": {}, 
        "replications": {}, 
        "roles": {
            "ansible": {
                "canlogin": true, 
                "member_of": [
                    "test"
                ], 
                "superuser": false, 
                "valid_until": ""
            }, 
        "settings": {
           ....,
          "work_mem": {
              "boot_val": "4096", 
              "context": "user", 
              "max_val": "2147483647", 
              "min_val": "64",
              "pretty_val": "4MB",
              "setting": "4096", 
              "sourcefile": "/var/lib/pgsql/11/data/postgresql.auto.conf", 
              "unit": "kB", 
              "val_in_bytes": 4194304, 
              "vartype": "integer",
              "pending_restart": true,
          }, 
           ....,
          "pending_restart_settings": [
            "port", 
            "shared_buffers"
           ],
    }, 
},
```

Andersson007 committed:
New module postgresql_facts: Gathers facts about remote PostgreSQL instances

Andersson007 committed:
New module postgresql_facts: fix default values of function

Andersson007 committed:
New module postgresql_facts: add extension support

Andersson007 committed:
New module postgresql_facts: fixed typo

Andersson007 committed:
New module postgresql_facts: added integration test

Andersson007 committed:
New module postgresql_facts: fix test error

Andersson007 committed:
New module postgresql_facts: add info about min server version

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 commented:
I have no idea why errors occurred related with queries to replication stat tables when I pushed the commit with integration tests.
On my local machine it was done in the docker container well..
```
test/runner/ansible-test integration --docker-no-pull --docker fedora28 -v postgresql >> ~/test.log
```
I added info in the module documentation that PostgreSQL must be 9.4 version or later (replication slots were added since this version).
@mattclay @dagwieers what must I do to pass this ansibot tests well?

dagwieers commented:
@Andersson007 You will have to look at the Shippable CI details and see why it failed for some of the test environments. This is the one for CentOS-6.
```
04:19 The full traceback is:
04:19 Traceback (most recent call last):
04:19   File "/tmp/ansible_postgresql_facts_payload_V_geHv/__main__.py", line 475, in __exec_sql
04:19     self.cursor.execute(query)
04:19   File "/usr/lib64/python2.6/site-packages/psycopg2/extras.py", line 118, in execute
04:19     return _cursor.execute(self, query, vars, async)
04:19 ProgrammingError: relation "pg_stat_replication" does not exist
04:19 LINE 1: ...ent_hostname, r.backend_start::text, r.state FROM pg_stat_re...
04:19                                                              ^
```
The problem is the same for CentOS-7, RHEL-7 and Ubuntu 14.04.

Andersson007 committed:
New module postgresql_facts: added repl stat tables checks

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 committed:
New module postgresql_facts: fix

Andersson007 commented:
ready_for_review

dagwieers commented:
Can the return facts be documented ? It really adds value if users can see what facts are returned.

For example see: https://docs.ansible.com/ansible/devel/modules/win_disk_facts_module.html#returned-facts

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+      - Limit collected subsets by comma separated list. Allowable values are C(version),
+        C(databases), C(settings), C(tablespaces), C(languages), C(roles), C(namespaces),
+        C(replications), C(repl_slots), C(extensions). By default, collects all subsets.
+        You can use shell-style (fnmatch) wildcard to pass groups of values (see Examples).
```
We tend to add parameters type for each parameter, so it ends up in the documentation.
```suggestion
        You can use shell-style (fnmatch) wildcard to pass groups of values (see Examples).
      type: list
```
Do this for all parameters.

Andersson007 replied:
This is a string that's split into code to a list. Is it necessary to implement this as a list?
Can I change it like this and remain 'str' type:
```
 27       - Limit collected subsets by comma separated names.
 28       - Allowable values are C(version),
 29         C(databases), C(settings), C(tablespaces), C(languages), C(roles), C(namespaces),
 30         C(replications), C(repl_slots), C(extensions). By default, collects all subsets.
 31       - You can use shell-style (fnmatch) wildcard to pass groups of values (see Examples).
```

Andersson007 replied:
Ok, I'll change them to lists. I've just seen your next suggestions.

dagwieers replied:
If you make it a list, and people provide a comma-separated string it will do the split for you.
So it will be backward compatible by default.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+    default: '*'
+  exclude_subset:
+    description:
+      - Exclude subsets from the collection. Allowable values are similar as in I(include_subset).
```
Please add separate items to the documentation so it is not one large blob.
```suggestion
      - Exclude subsets from the collection. 
      - Allowable values are similar as in I(include_subset).
```

Andersson007 replied:
Good notice

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+DOCUMENTATION = '''
+---
+module: postgresql_facts
+short_description: Gathers facts about remote PostgreSQL servers.
```
The short description does not need a trailing dot (all other descriptions do).
```suggestion
short_description: Gather facts about remote PostgreSQL servers
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+  exclude_subset:
+    description:
+      - Exclude subsets from the collection. Allowable values are similar as in I(include_subset).
+        You can also use shell-style (fnmatch) wildcard. Mutually exclusive with include_subset.
```
```suggestion
      - You can also use shell-style (fnmatch) wildcard.
      - Mutually exclusive with include_subset.
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+     systems, install the postgresql, libpq-dev, and python-psycopg2 packages
+     on the remote host before using this module.
+requirements: [ psycopg2 ]
+author: "Andrew Klychkov (@Andersson007)"
```
No need for quotes, make it a list.
```suggestion
author: 
- Andrew Klychkov (@Andersson007)
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+def main():
+    argument_spec = pgutils.postgres_common_argument_spec()
+    argument_spec.update(dict(
+        db=dict(default=''),
```
Please add explicit types here too.
```suggestion
        db=dict(type='str', default=''),
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+    # Do job:
+    pg_facts = PgClusterFacts(module, cursor)
+
+    if incl_subsets != "*":
```
~Please make `incl_subsets` and `excl_subsets` a list so that it supports comma-separated as well as accepts lists.~ This should actually become a single parameter named `gather_subet`.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+        else:
+            db_connection.commit()
+
+    kw['changed'] = changed
```
A facts module must never induce a change ! It should only gather information.

Andersson007 replied:
Yes. Must it have 'changed' == False or it must not have 'changed' key at all?

dagwieers replied:
Well the default is changed=False anyway, so there's no need, but if you do it anyway, there's no problem. The most important thing is that the module should not cause any changes on the system.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+        kw["host"] = module.params["login_unix_socket"]
+
+    if psycopg2.__version__ < '2.4.3' and sslrootcert is not None:
+        module.fail_json(
```
Make this one line
```suggestion
        module.fail_json(msg='psycopg2 must be at least 2.4.3 in order to user the ssl_rootcert parameter')
```

Andersson007 replied:
done

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+
+    if psycopg2.__version__ < '2.4.3' and sslrootcert is not None:
+        module.fail_json(
+            msg='psycopg2 must be at least 2.4.3 in order to user the ssl_rootcert parameter')
```
```suggestion
```

Andersson007 replied:
done

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,597 @@
+    try:
+        db_connection = psycopg2.connect(**kw)
+
+        cursor = db_connection.cursor(
```
Same here, there's no real need to have this on 2 lines...

Andersson007 replied:
I agree, this is legacy

Andersson007 commented:
@dagwieers , it sounds really good

Andersson007 committed:
New module postgresql_facts: ansible_facts -> postgresql_facts

Andersson007 committed:
New module postgresql_facts: fix ci

New module postgresql_facts: revision for review

New module postgresql_facts: revision for review, fix

dagwieers reviewed (changes requested):
A bunch of remarks, nothing very critical.

Andersson007 committed:
fix doc

Andersson007 commented:
@dagwieers , done for all of your notices.

It was difficult but I've got a lot of valuable experience to implement this ;)
FYI: Unfortunately, I needed to make force push.

ready_for_review

Andersson007 commented:
ready_for_review

Andersson007 committed:
New module postgresql_facts: add data types to param dict

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+    type: list
+  db:
+    description:
+      - Name of database to connect.
```
```suggestion
      - Name of database to connect.
    type: str
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+      - Name of database to connect.
+  port:
+    description:
+      - Database port to connect.
```
```suggestion
      - Database port to connect.
    type: int
```
Please do this for all the parameters.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+def main():
+    argument_spec = pgutils.postgres_common_argument_spec()
+    argument_spec.update(dict(
+        db=dict(default='', type=str),
```
Beware, it needs to be `type='str'`. Please always put the `type` option at the start if your dict, it's easier to read the argspec (as every parameter must have a type-definition).
Usually we do not set defaults so you can easily detect if the value was set, or not (None).
```suggestion
        db=dict(type='str'),
```

dagwieers commented:
So regarding the `incl_subset` and `excl_subset` the standard way of doing this in facts-modules is to have a `gather_subset` parameter that includes the subsets you want to include, and it defaults to `all`.

So this would replace the other two parameters by a single one. Recursively grep for `gather_subset` in your modules-directory and you'll see how other modules do this.

dagwieers commented on test/integration/targets/postgresql/tasks/postgresql_facts.yml:
```diff
@@ -0,0 +1,101 @@
+# Copyright: (c) 2019, Andrew Klychkov (@Andersson007) <aaklychkov@mail.ru>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+- name: "postgresql_facts - test return values"
```
So not something you have to change, but I noticed this playbook is quoting a lot of things that do not require quoting at all. For documentation and examples we recommend to only quote where it is required by YAML to quote.
This helps users to understand when quotes are needed.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+        excl_subset=dict(default="", type=str),
+        ssl_mode=dict(default='prefer', choices=[
+            'disable', 'allow', 'prefer', 'require', 'verify-ca', 'verify-full']),
+        ssl_rootcert=dict(default=None, type=str),
```
You don't need to add `default=None` as it is implied if there is no default.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+    ))
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True
```
We tend to add commas after every entry, even if it is the last one.
So that if you extend a dictionary or a list, you do not have to modify the previous line.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+                return res
+        except SQLParseError as e:
+            self.module.fail_json(msg=to_native(e),
+                                  exception=traceback.format_exc())
```
You don't have to do this explicitly, Ansible will add this for you automatically.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+except ImportError:
+    HAS_PSYCOPG2 = False
+else:
+    HAS_PSYCOPG2 = True
```
Why not simply do:
```python
try:
    import psycopg2
    import psycopg2.extras
    HAS_PSYCOPG2 = True
except ImportError:
    HAS_PSYCOPG2 = False
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+
+def main():
+    argument_spec = pgutils.postgres_common_argument_spec()
+    argument_spec.update(dict(
```
You can make this easier:
```suggestion
    argument_spec.update(
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+    # Rollback transaction, if checkmode.
+    # Otherwise, commit transaction:
+    # (It doesn't make sense in this praticular case, just for order)
+    if module.check_mode:
```
Please get rid of this whole block. We should be committing or rolling back anything, it is a facts-module.

dagwieers commented on test/integration/targets/postgresql/tasks/postgresql_facts.yml:
```diff
@@ -0,0 +1,101 @@
+      - ver*
+      - rol*
+  register: result
+  ignore_errors: True
```
Just FYI, for documentation and examples we prefer booleans as yes/no, but we do also allow true/false.
We do not want to promote True/False or other YAML 1.1 boolean values to keep the documentation and examples consistent.

For integration tests, we do not enforce any standard.

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+            if i[2]:
+                set_info["unit"] = i[2]
+            else:
+                set_info["unit"] = ""
```
In Python you can do this, which could make your code a bit more concise.
```python
                set_info['unit'] = i[1] if i[1] else ''
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+            type: dict
+            sample: 47823
+            contains:
+                usename:
```
Should this not be **username** ?

Andersson007 replied:
This is because in the source system table pg_stat_replication, this column named "usename". I tried to name parameters in accordance to their names in postgres statistics tables.
Well.. I was thinking about it too. The question is what's better in Ansible case... maybe this module will be used not only DBAs. I should think. Maybe would be better to name it like **username**. I don't really understand why postgresql developers somewhere named similar column in system tables/views like "rolname", "rolename", "usename", etc. ;)

Andersson007 replied:
According to my explanation above, give me your opinion, in this case (Ansible) **username** is better?

Andersson007 replied:
I'll change because it looks like a typo ;)

dagwieers replied:
If it is a low-level thing, by all means keep it at `usename` !

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+        for i in res:
+            repl_pid = i[0]
+            repl_info = {}
+            repl_info["usename"] = i[1]
```
Should this not be **username** ?
Why not create the dict with some values in it ?
```suggestion
            repl_info = dict(username=i[1])
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,916 @@
+        db_dict = {}
+        for i in res:
+            db_name = i[0]
+            db_info = {}
```
You can replace this whole block using a single assignment:
```python
                db_dict[db_name] = dict(
                    owner=i[1],
                    encoding=i[2],
                    collate=i[3],
                    ctype=i[4],
                    access_priv=i[5] if i[5] else '',
                    size=i[6],
                )
```

Andersson007 replied:
this is the best hint I've recently viewed ;)

dagwieers reviewed (changes requested):
Looking good, some more feedback included.

Andersson007 committed:
New module postgresql_facts: code refactoring

Andersson007 committed:
# This is a combination of 3 commits.
# The first commit's message is:
New module postgresql_facts: incl_subset => filter

# This is the 2nd commit message:

New module postgresql_facts: fix a typo

# This is the 3rd commit message:

New module postgresql_facts: fixes

Andersson007 commented:
@dagwieers , seems I've done changes. The key points are:
1. I understood that there aren't any subsets in postgresql_facts...
gather_subset doesn't make sense in this module because the full set of facts quite small.
In the meaning of like in https://docs.ansible.com/ansible/latest/modules/setup_module.html 
incl_subsets / excl_subsets was just a filter (like the "filter=someparam*" in setup module).
I changed it to "filter=". See examples in the topic.
2. I formatted it as you requested.
3. I left 'usename' as it was.

Andersson007 commented:
ready_for_review

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,878 @@
+          description: Current value of the parameter.
+          returned: always
+          type: str
+          sample: 49152
```
We prefer that returned information is in bytes here. So that values can be converted without complexity.

Andersson007 replied:
@dagwieers I'm afraid the implementation of this suggestion makes the code very complex because a parameter can have a value like in following output from PostgreSQL system view pg_settings:
```
                name                 |                setting                 | unit 
-------------------------------------+----------------------------------------+------
 allow_system_table_mods             | off                                    | 
 application_name                    | psql                                   | 
 archive_command                     | (disabled)                             | 
 archive_mode                        | off                                    | 
 archive_timeout                     | 0                                      | s
 array_nulls                         | on                                     | 
 authentication_timeout              | 60                                     | s
 autovacuum                          | on                                     | 
 autovacuum_analyze_scale_factor     | 0.1                                    | 
 autovacuum_analyze_threshold        | 50                                     | 
 autovacuum_freeze_max_age           | 200000000                              | 
 autovacuum_max_workers              | 3                                      | 
 autovacuum_multixact_freeze_max_age | 400000000                              | 
 autovacuum_naptime                  | 60                                     | s
 autovacuum_vacuum_cost_delay        | 20                                     | ms
 autovacuum_vacuum_cost_limit        | -1                                     | 
 autovacuum_vacuum_scale_factor      | 0.2                                    | 
 autovacuum_vacuum_threshold         | 50                                     | 
 autovacuum_work_mem                 | -1                                     | kB
 backend_flush_after                 | 0                                      | 8kB
 backslash_quote                     | safe_encoding                          | 
 bgwriter_delay                      | 200                                    | ms
 bgwriter_flush_after                | 64                                     | 8kB
 bgwriter_lru_maxpages               | 100                                    | 
 bgwriter_lru_multiplier             | 2                                      | 
 block_size                          | 8192                                   | 
 bonjour                             | off                                    | 
 bonjour_name                        |                                        | 
 bytea_output                        | hex                                    | 
 check_function_bodies               | on                                     | 
 checkpoint_completion_target        | 0.5                                    | 
 checkpoint_flush_after              | 32                                     | 8kB
 checkpoint_timeout                  | 300                                    | s
 checkpoint_warning                  | 30                                     | s
 client_encoding                     | UTF8                                   | 
 client_min_messages                 | notice                                 | 
 cluster_name                        |                                        | 
 commit_delay                        | 0                                      | 
 commit_siblings                     | 5                                      | 
 config_file                         | /var/lib/pgsql/10/data/postgresql.conf | 
```
And this is a small part... there are other cases like 'logical', 'hot standby', etc

dagwieers replied:
Remember that you won't be able to change this in the future, because people depend on this design.

Andersson007 replied:
I added an additional value val_in_bytes that will be returned if supported:
```
        "bgwriter_flush_after": {
            "boot_val": "64", 
            "context": "sighup", 
            "max_val": "256", 
            "min_val": "0", 
            "setting": "64", 
            "sourcefile": "", 
            "unit": "8kB", 
            "val_in_bytes": 524288, 
            "vartype": "integer"
        }, 
```

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,878 @@
+        U(https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-STATS-VIEWS-TABLE).
+      returned: if pg_stat_replication view existent
+      type: dict
+      sample: 47823
```
This seems weird.

Andersson007 replied:
This is an example of a possible key of the dict. The key is a pid of a replication process that allows us to uniq it in a postgres instance. I'll try to change it to the dict

Andersson007 replied:
Changed

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,878 @@
+              description: Database size.
+              returned: always
+              type: str
+              sample: 7861 kB
```
This ought to be in bytes as well. So that these values can be used without complexity (i.e. being parsed).

Andersson007 replied:
Changed  to bytes

dagwieers reviewed (changes requested):
After reviewing again, I found that some values returned are not in bytes, and are returned as string.
We need to ensure that values are using the base unit (here bytes).

Andersson007 commented:
@dagwieers , seems I fixed all conflicts  with your commit but it failed. I don't really know how solve this situation properly. Force push?

Andersson007 commented:
Oh, Ok, sorry, I've just seen the message from ansibot

Andersson007 committed:
New module postgresql_facts: incl_subset => filter

New module postgresql_facts: fix a typo

New module postgresql_facts: fixes

Various cosmetic and doc changes

Andersson007 commented:
I'm sorry, I didn't see your last commit and change the file that you changed before. So I did rebase after conflicts.

Andersson007 commented:
ready_for_review

Andersson007 commented:
cc @sebasmannem @amenonsen

Andersson007 committed:
New module postgresql_facts: fix filter type

Andersson007 committed:
New module postgresql_facts: fix sanity

dagwieers commented:
So we recently had a discussion about this for another module, and the problem here is that we are using a common namespace for specific information about postgresql databases. That is an issue.

IMO These facts-modules should not be using `ansible_facts` because they do not expose information about the host/target. I think it would be better if the user has to register this information rather than updating the facts namespace.

Andersson007 committed:
New module postgresql_facts: change ansible_facts to postgresql_facts

Andersson007 commented:
@dagwieers I actually thought about this too when I was writing the module. I've changed ansible_facts to postgresql_facts. Is it OK?

Andersson007 committed:
New module postgresql_facts: fix tests

Andersson007 commented:
ready_for_review

dagwieers commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,905 @@
+    pg_facts = PgClusterFacts(module, cursor)
+
+    if filter_:
+        kw['postgresql_facts'] = pg_facts.collect(filter_)
```
Why put this in an additional `postgresql_facts` layer ? It only makes it harder for people to get at that information.
Since they already name the `register:` variable themselves, there's no need to add another namespace here.

(Look at the examples, the `postgredql_facts` key in the datastructure is irrelevant, it does not add anything to the user, it only makes it harder to get at the information they need.

dagwieers replied:
This whole block at the end can be reduced to:
```python
    module.exit_json(pg_facts.collect(filter_))
```

Andersson007 replied:
great notice! changed

Andersson007 committed:
New module postgresql_facts: fixes

Andersson007 committed:
New module postgresql_facts: fixes

Andersson007 committed:
New module postgresql_facts: fixes

Andersson007 commented:
ready_for_review

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+  contains:
+    nspacl:
+      description:
+      - Assess privileges U(https://www.postgresql.org/docs/current/catalog-pg-namespace.html).
```
I think you mean 'Access privilleges' instead of 'Assess privilleges'

Andersson007 replied:
!

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+DOCUMENTATION = r'''
+---
+module: postgresql_facts
+short_description: Gather facts about remote PostgreSQL servers
```
I would change 'remote' into 'remote and local'.
Local works too. And it is even the default.

Andersson007 replied:
I'll remove 'remote' at all

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+module: postgresql_facts
+short_description: Gather facts about remote PostgreSQL servers
+description:
+- Gathers facts about remote PostgreSQL servers.
```
I would change 'remote' into 'remote and local'.
Local works too. And it is even the default.

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+    description:
+    - Host running PostgreSQL.
+    type: str
+  login_unix_socket:
```
This is an extra option with very little functionality.
I would leave this extra option out and document that you can set login_host to the socket dir too, in the Notes section.

Andersson007 replied:
I'd prefer to remain it for compatibility

kustodian replied:
Compatibility with what?

Andersson007 replied:
Maybe somebody use this option. It is presented in other PG modules.
If you absolutely sure that it's useless, OK, I'll remove it.
Are you sure? ;)

kustodian replied:
Sorry, I had no idea old PG modules also had that option, because in PG clients `hosts` parameter was always set to the path of the unix socket. I guess it's better to leave it, and maybe in the next version we can make it an alias to host or something similar for all modules.

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+  become: yes
+  become_user: postgres
+  postgresql_facts:
+    filter: ver*,ext*
```
I think it would be better to quote:
    filter: "ver*,ext*"

Andersson007 replied:
I must use quotes in the example
```
- name: Collect all subsets, excluding settings and roles
  postgresql_facts:
    filter: "!settings,!roles"
```
because "!" doesn't work without it in this particular example.
Generally, better to avoid quotes in the examples where not necessarily because it's simplier and works.

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+          returned: always
+          type: str
+          sample: 8189415
+extensions:
```
What I find confusing, is that we have two levels of info crossing here.
Basically, this info is only for the database you are connecting to.
But it is listed on the global level, next to the databases, so it looks as if it where global.

sebasmannem replied:
I think the nicest way would be to login to all databases and collect this per database.
You could then set this per database on the databases entry level.
And then this entry level should better be info from pg_available_extensions, with the addition of a list of databases it is installed in.

sebasmannem replied:
Going through all databases is (a little) more intrusive, but it fits to what I expect of a facts module (in that it delivers all facts, not just for one db). You could even switch this on/off, by setting db to '*' or to a (list of) specific database(s), but then '*' should be the default.

Andersson007 replied:
Good idea

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+    - If you pass including and excluding values to the filter, for example, I(filter=!settings,ver),
+      the excluding values will be ignored.
+    type: list
+  db:
```
This should rather be login_db

Andersson007 replied:
Sounds good.
In all of the existing modules they're used db, port, etc. So, I'll add aliases: login_db, login_port,...

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+    description:
+    - Name of database to connect.
+    type: str
+  port:
```
This should rather be login_port

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+      returned: always
+      type: str
+      sample: pg_catalog
+languages:
```
What I find confusing, is that we have two levels of info crossing here.
Basically, this info is only for the database you are connecting to.
But it is listed on the global level, next to the databases, so it looks as if it where global.

sebasmannem replied:
I would propose same fix as the previous one (on extensions), but there is no pg_available_languages, so I would then skip this level entirely, or build it from info scanned from all databases, and adding a list of databases it was installed in.

Andersson007 replied:
Very good notice! I'll think about it.

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+      returned: always
+      type: str
+      sample: postgres
+namespaces:
```
What I find confusing, is that we have two levels of info crossing here.
Basically, this info is only for the database you are connecting to.
But it is listed on the global level, next to the databases, so it looks as if it where global.

sebasmannem replied:
I would propose same fix as on extensions, but there is no pg_available_namespaces, so I would then skip this level entirely, or build it from info scanned from all databases, and adding a list of databases it was created in.

Andersson007 replied:
Very good notice! I'll think about it.

sebasmannem commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,921 @@
+  sample: { "slot0": { "active": false, "database": null, "plugin": null, "slot_type": "physical" } }
+  contains:
+    active:
+      description: True if this slot is currently actively being used.
```
I know this is the exact documentation from postgresl.org, but I think this 'sounds' as if a receiver is currently connected to it. Maybe we should add some info, as that 'True means that a receiver has connected to it, and it is currently reserving archives', or something like that.

sebasmannem reviewed (changes requested):
I have gone through the documentational part already (DOCUMENTATION, EXAMPLE, RETURN), and these are the comments for those. I will review the python code later.

Andersson007 committed:
New module postgresql_facts: doc fixes

Andersson007 commented:
@sebasmannem , seems all of your requests are satisfied.
I added getting info about namespaces, ext, and langs for each databases. So, now return value looks like this:
```
    "databases": {
        "acme": {
            "access_priv": "=Tc/postgres\npostgres=CTc/postgres\nalice=c/postgres", 
            "collate": "en_US.UTF-8", 
            "ctype": "en_US.UTF-8", 
            "encoding": "UTF8", 
            "extensions": {
                "hstore": {
                    "description": "data type for storing sets of (key, value) pairs", 
                    "extversion": {
                        "major": 1, 
                        "minor": 5
                    }, 
                    "nspname": "public"
                }, 
                "plpgsql": {
                    "description": "PL/pgSQL procedural language", 
                    "extversion": {
                        "major": 1, 
                        "minor": 0
                    }, 
                    "nspname": "pg_catalog"
                }
            }, 
            "languages": {
                "c": {
                    "lanacl": "", 
                    "lanowner": "postgres"
                }, 
                "internal": {
                    "lanacl": "", 
                    "lanowner": "postgres"
                }, 
                "plpgsql": {
                    "lanacl": "", 
                    "lanowner": "postgres"
                }, 
                "sql": {
                    "lanacl": "", 
                    "lanowner": "postgres"
                }
            }, 
            "namespaces": {
                "information_schema": {
                    "nspacl": "{postgres=UC/postgres,=U/postgres}", 
                    "nspowner": "postgres"
                }, 
                "pg_catalog": {
                    "nspacl": "{postgres=UC/postgres,=U/postgres}", 
                    "nspowner": "postgres"
                }, 
                "pg_temp_1": {
                    "nspacl": "", 
                    "nspowner": "postgres"
                }, 
                "pg_toast": {
                    "nspacl": "", 
                    "nspowner": "postgres"
                }, 
                "pg_toast_temp_1": {
                    "nspacl": "", 
                    "nspowner": "postgres"
                }, 
                "public": {
                    "nspacl": "{postgres=UC/postgres,=UC/postgres}", 
                    "nspowner": "postgres"
                }
            }, 
            "owner": "postgres", 
            "size": "8293375"
        }, 
```

Andersson007 commented:
ready_for_review

jctanner commented:
!waffling community_review

Andersson007 commented:
@sebasmannem , waiting for you, PTAL

kustodian commented:
I played with the module a little bit and it looks very useful I would already have a use case for it. One thing that I miss in the module is for settings that it also returns a "pretty value" of the setting that is returned by `SHOW`. The use case for this is if I want to compare if the value of some setting is the same as in a var, e.g. if I set a var `pg_shared_buffers: 128MB` and if I want to compare my var value with the value that is returned with this module, it's not an easy task.

Andersson007 committed:
New module postgresql_facts: added pretty_val

Andersson007 commented:
@kustodian , very good idea, added pretty_val
```
        "work_mem": {
            "boot_val": "4096", 
            "context": "user", 
            "max_val": "2147483647", 
            "min_val": "64", 
            "pretty_val": "4MB", 
            "setting": "4096", 
            "sourcefile": "/var/lib/pgsql/11/data/postgresql.auto.conf", 
            "unit": "kB", 
            "val_in_bytes": 4194304, 
            "vartype": "integer"
        }, 
```
waiting for your feedback

kustodian commented:
@Andersson007 what about adding `pending_restart`. It's very easy to return that field as well (I guess there would need to be a check for PG version). I added `pending_restart` to my local version of the module and it's not trivial to extract if one of the values has this set to true (using Jinja2). Because of this I was thinking that maybe the module could return and additional field with a list of items that are pending a restart, or something similar. For example `pending_restart_settings: [ 'shared_buffers', 'port' ]`. This way just checking if `pending_restart_settings != []` would be enough to know if PG needs to be restarted.

The reason why I'm mentioning this use case is because I already have it and probably anyone who is configuring PG with Ansible. Currently we solve it by having a list of settings that require a PG restart, and after we do a reload, we compare these values with current values, if they are different, that means we need a restart, if they are the same, reload was enough. The problem with this is that we have to keep track of the settings that require a restart and it's kind of slow because we have to check these values using a `loop`. So this module looks to perfectly improve this use case.

Andersson007 committed:
New module postgresql_facts: added pending restart

Andersson007 commented:
@kustodian , good note again! Added 
```
    "pending_restart_settings": [
        "port", 
        "shared_buffers"
    ], 
```
and
```
        "shared_buffers": {
            "boot_val": "1024", 
            "context": "postmaster", 
            "max_val": "1073741823", 
            "min_val": "16", 
            "pending_restart": true, 
            "pretty_val": "128MB", 
            "setting": "16384", 
            "sourcefile": "/var/lib/pgsql/11/data/postgresql.conf", 
            "unit": "8kB", 
            "val_in_bytes": 134217728, 
            "vartype": "integer"
        }, 
```
waiting for your feedback

kustodian commented:
I have nothing to add or remove. This looks good to me.

shipit

Andersson007 commented:
@kustodian , thank you much for this valuable review

Andersson007 committed:
New module postgresql_facts: fix documentation

Andersson007 committed:
New module postgresql_facts: fix documentation

Andersson007 commented:
@kostiantyn-nemchenko , could you please look at this PR and approve or give us notes to make it better?
(would be nice to see it and 51875 merged before 21 March 2.8 feature freeze, seems it'll be very useful for all)

kostiantyn-nemchenko commented:
@Andersson007 Yes, I will try to find some time today/tomorrow for it.

Andersson007 commented:
@kostiantyn-nemchenko , thank you much, waiting for your feedback

cmfrtl commented:
shipit

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,996 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2018, Andrew Klychkov (@Andersson007) <aaklychkov@mail.ru>
```
2019

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,996 @@
+                type: int
+                sample: 0
+            nspname:
+              description: Namecpase where the extension is.
```
Namespace

Andersson007 replied:
very funny double typo in one word ;)

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,996 @@
+    filter: "!settings,!roles"
+
+# On FreeBSD with PostgreSQL 9.5 version and lower use pgsql user to become
+# and pass "postgresq" as a database to connect to
```
"postgres"?

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,996 @@
+      sample: acme
+    plugin:
+      description:
+      - The base name of the shared object containing the the output plugin
```
the

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,996 @@
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = r'''
```
No session_role parameter support

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,996 @@
+  become_user: postgres
+  postgresql_facts:
+    filter:
+    - "!settings"
```
Example with `!settings` is already present on line 122. Would be better to not repeat the same facts

kostiantyn-nemchenko commented on lib/ansible/modules/database/postgresql/postgresql_facts.py:
```diff
@@ -0,0 +1,996 @@
+                 "ELSE 'No Access' END, "
+                 "t.spcname "
+                 "FROM pg_catalog.pg_database AS d "
+                 "JOIN pg_catalog.pg_tablespace t on d.dattablespace = t.oid")
```
Should we exclude template databases here?

jctanner commented:
!waffling community_review

Andersson007 committed:
New module postgresql_facts: fixes by KN

Andersson007 committed:
New module postgresql_facts: fixed sanity

Andersson007 committed:
New module postgresql_facts: fixed tests

Andersson007 commented:
@kostiantyn-nemchenko , you are a guru of review ;) thanks !
fixed all issues. If you are saticfied, please, approve it and shipit

ready_for_review

kostiantyn-nemchenko commented:
shipit

Andersson007 commented:
thanks all ! waiting for 2019-05-02 Release

gundalow commented:
@Andersson007 Thank you for this new module
@dagwieers @kustodian @kostiantyn-nemchenko @sebasmannem Thank you for the reviews.

Merged into `devel` will be released in Ansible 2.8
